rfc/indexing-design.md

```diff
@@ -0,0 +1,228 @@
+
+The blob-index is stored in the user's space along with the CAR file(s) containing the blob data. This is done in the [w3 add-blob operation](https://github.com/w3s-project/specs/blob/main/w3-blob.md#add-blob). The w3s service creates [location commitments](https://github.com/w3s-project/specs/blob/main/w3-blob.md#location-commitment) for the blob shards and for the blob-index. 
+
+After the add-blob process completes, and the user receives an [accept-blob receipt](https://github.com/w3s-project/specs/blob/main/w3-blob.md#accept-blob-receipt), the user may then choose to make their data publicly queryable by publishing it on [IPNI](https://github.com/ipni/specs/blob/main/IPNI.md). The user optionally invokes an [index-add](https://github.com/w3s-project/specs/blob/main/w3-index.md#index-add) capability to publish the blob-index multihashes to IPNI so that they can be used to look up location commitments and retrieve blob-index information and blob data. See [W3 Index](https://github.com/w3s-project/specs/blob/main/w3-index.md#w3-index) for more.
+
+After publishing to IPNI, any user looking for content can query IPNI for the blob-index of that content. The user read the blob-index to determine what data shard needs to be retrieved, and then asks IPNI for the location of that shard.
+
+Finally the user uses the shard location, with range information from the blob index, to retrieve a specific piece of data from w3s.
+
+## Index Query
+
+A w3s client can query indexed information by specific attributes, and can filter retrieved query results.
+
+The queryable attributes are as follows.
+- Site: the location(s) where all blob data is available
+- Shard: the location(s) where a specific shard of data is available
+- Slice: the location(s) within a shard where a block of data is available
+
+Query results can be filtered to select only results that the user has access to, or for data supplied by specific data providers.
+
+### Query Specification
+
+```
+match = {
+  subject=[myID, myOrgID],
+  location=spBaseURL,
+}
+results = Query(multihash="bafy...", match)
+```
+
+### Query Operation Diagram
+
+![w3up-query-diagram](w3up-query-diagram.png)
+
+## Indexing Components
+
+The indexing components are intended for deployment on the w3up gateway. They may also be deployed to the w3up client, where they will not be shared with other clients.
+
+### W3up Index
+
+The w3up index layer is responsible for reading a query specification, and then
+1. Creating a set of IPNI queries needed to get the requested index content
+2. Sending the queries to the w3up IPNI cache
+3. Filtering returned results
+4. Presenting results to the query client.
+
+### W3up IPNI Cache
+
+The w3up IPNI Cache is a cache that holds IPNI query results. The cache is:
```

Comment:
So this is my most substantive comment:

Is this an IPNI cache? Or is it a IPNI + sharded-dag-index cache?

As I understand it, IPNI results will ONLY contain a location to retrieve a Sharded Dag Index.

You'll need to read that sharded-dag-index / blob-index seperately (sidebar: I'm finding the distinction very difficult to understand -- can we have a terms section)

So are we caching that? I think the answer is yes, cause otherwise every query contains an extra round trip. I imagine these sharded-dag-indexes are evicted more often than the IPNI queries, since they're going to be larger. But I think to support a shard/slice query without blocking on an additional round trip, we should be caching these.

And if we're caching indexes, I think we should be explicit and not call this an IPNI cache.